### PR TITLE
Add an MVP implementation of miner-side discovery mode

### PR DESCRIPTION
**Pull request by JayKickliter**
This PR adds an MVP implementation for miner-side discovery mode.

Outstanding questions:

- what to do when router requests miner to enter a period of discovery mode but the previous batch of uplinks is not complete?
- what to do when `miner_discovery_worker` dies?

### CODE REVIEW

**Comment by macpie** [comment]
> what to do when router requests miner to enter a period of discovery mode but the previous batch of uplinks is not complete?

I would think we should then ignore it to avoid spam

> what to do when `miner_discovery_worker` dies?

If you mean normally I would do nothing, its job is done.

**Comment by macpie** [comment]
For the ignore, as your process is named a simple `is_process_alive` should tell you if it is still running.

File: src/handlers/miner_discovery_handler.erl [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+            lager:info("failed to verify signature for ~p (sig=~p)", [
+                blockchain_utils:addr2name(HostpostPubKeyBin),
+                Sig
+            ]);
+        true ->
+            case miner_lora:location_ok() of
+                true ->
+                    miner_discovery_worker:start(Packets);
+                false ->
+                    lager:warning("can not start discovery mode until location is ok"),
```
Comment (michaeldjeffrey) [inline src/handlers/miner_discovery_handler.erl:77]: Do Miner's have a way of sending error messages back? So we don't have to burrow to see if something might not be happening?

File: src/miner.erl [hunk 2]
```diff
@@ -310,6 +310,8 @@ version() ->
 init(_Args) ->
     lager:info("STARTING UP MINER"),
     ok = blockchain_event:add_handler(self()),
+    %% TODO: Maybe put this somewhere else?
+    ok = miner_discovery_handler:add_stream_handler(blockchain_swarm:tid()),
```
Comment (Vagabond) [inline src/miner.erl:314]: Add it to the supervision tree

File: src/miner_discovery_worker.erl [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    timer:send_after(?DEFAULT_TRANSMIT_DELAY_MS, self(), tick),
+    {noreply, State#state{packets = Packets}};
+handle_info(_Info, State) ->
+    {noreply, State}.
+
+%% ------------------------------------------------------------------
+%% Internal
+%% ------------------------------------------------------------------
+
+%% TODO: taken from miner_onion_server, refactor to common module?
```
Comment (vihu) [inline src/miner_discovery_worker.erl:94]: It's fine for now, we should be able to lookup tx_power and spreading using the location of the hotspot once some of the PoC related work lands.

File: src/handlers/miner_discovery_handler.erl [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+handle_info(_Type, _Msg, State) ->
+    lager:warning("test ~p got info ~p", [_Type, _Msg]),
+    {noreply, State}.
+
+%% ------------------------------------------------------------------
+%% Internal Function Definitions
+%% ------------------------------------------------------------------
+
+-spec verify_signature(
+    Hostpost :: binary(),
```
Comment (vihu) [inline src/handlers/miner_discovery_handler.erl:98]: Hostpost -> Hotspot
HostpostPubkeyBin -> HotspotPubkeyBin

:)

File: src/handlers/miner_discovery_handler.erl [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    {ok, #state{}}.
+
+handle_data(server, Data, State) ->
+    lager:info("got ~p ~p", [Data, State]),
+    #discovery_start_pb{
+        hotspot = HotspotPubkeyBin,
+        packets = Packets,
+        signature = Sig
+    } = discovery_pb:decode_msg(Data, discovery_start_pb),
+    Hotspot = erlang:list_to_binary(libp2p_crypto:bin_to_b58(HotspotPubkeyBin)),
```
Comment (vihu) [inline src/handlers/miner_discovery_handler.erl:65]: Seems a little odd that we're converting a binary to a binary again? If there's a particular reason to do so, maybe add a comment here?

**Review by evanmcc — APPROVED** [review]

**Pull request merged**